### DISCUSSION TITLE
Given lat/lon/alt/bearing/pitch/velocity, how do you get linear velocity for the Entity_State_PDU?

### DISCUSSION THREAD

**Discussion by Skhmt** [karljj1/kdis#33] [Q&A]
I have lat, lon, altitude, bearing, pitch, roll, and forward velocity. Assuming an entity that is just moving straight with dead reckoning formula 2 (FPW, world-centered instead of body-centered), how do I calculate the Entity_State_PDU's EntityLinearVelocity?

I couldn't find any utilities for doing such and the example just sets them to 0, 0, 0.

**Comment by karljj1** [comment]
Hi,
It has been a long time since I did any of this so I can't be certain but I think it was created by converting the position into geocentric cartesian coordinates and then calculating a vector velocity. 
So you could use `GeodeticToGeocentric` and maybe also `HeadingPitchRollToEuler`, then you just need to know the linear velocity based on these values.

**Comment by Skhmt** [comment]
It's funny after reading your response and working on this for a week and some change, I thought of an answer while driving in to work. I basically just use vicenty's algorithm to translate a lat/lon to a new lat/lon given distance and bearing, then get geocentric ECEF coords for both and do simple subtraction for the difference to get the velocity vector. Tested and it seems to work. Thanks!
